src/handler/sign_tx.c

```diff
@@ -1,103 +1,423 @@
-/*****************************************************************************
- *   Ledger App Boilerplate.
- *   (c) 2020 Ledger SAS.
- *
- *  Licensed under the Apache License, Version 2.0 (the "License");
- *  you may not use this file except in compliance with the License.
- *  You may obtain a copy of the License at
- *
- *      http://www.apache.org/licenses/LICENSE-2.0
- *
- *  Unless required by applicable law or agreed to in writing, software
- *  distributed under the License is distributed on an "AS IS" BASIS,
- *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- *  See the License for the specific language governing permissions and
- *  limitations under the License.
- *****************************************************************************/
-
 #include <stdint.h>   // uint*_t
 #include <stdbool.h>  // bool
 #include <stddef.h>   // size_t
-#include <string.h>   // memset, explicit_bzero
+#include <string.h>   // explicit_bzero, memmove
 
 #include "os.h"
 #include "cx.h"
 
 #include "sign_tx.h"
+
+#include "../hathor.h"
 #include "../sw.h"
 #include "../globals.h"
-#include "../crypto.h"
 #include "../ui/display.h"
+#include "../ui/menu.h"
 #include "../common/buffer.h"
+#include "../common/read.h"
 #include "../transaction/types.h"
 #include "../transaction/deserialize.h"
 
-int handler_sign_tx(buffer_t *cdata, uint8_t chunk, bool more) {
-    if (chunk == 0) {  // first APDU, parse BIP32 path
-        explicit_bzero(&G_context, sizeof(G_context));
-        G_context.req_type = CONFIRM_TRANSACTION;
-        G_context.state = STATE_NONE;
-
-        if (!buffer_read_u8(cdata, &G_context.bip32_path_len) ||
-            !buffer_read_bip32_path(cdata,
-                                    G_context.bip32_path,
-                                    (size_t) G_context.bip32_path_len)) {
-            return io_send_sw(SW_WRONG_DATA_LENGTH);
+
+bool verify_change_output(tx_output_t output, uint32_t *bip32_path, uint8_t bip32_len) {
+    uint8_t hash[20];
+    cx_ecfp_public_key_t public_key;
+    cx_ecfp_private_key_t private_key;
+    uint8_t chain_code[32];
+
+    derive_private_key(&private_key, chain_code, bip32_path, bip32_len);
+    init_public_key(&private_key, &public_key);
+    compress_public_key(public_key.W);
+    hash160(public_key.W, 33, hash);
+
+    // erase data
+    explicit_bzero(&private_key, sizeof(private_key));
+    explicit_bzero(&public_key, sizeof(public_key));
+    explicit_bzero(&chain_code, sizeof(chain_code));
+
+    // 0 means equals
+    return memcmp(hash, output.pubkey_hash, 20) == 0;
+}
+
+void read_change_output_info(buffer_t *cdata) {
+    uint8_t buffer[MAX_BIP32_PATH * 4 + 1] = {0};
+    uint8_t tmp;
+
+    // 1 byte for has_change_output and bip32 path len
+    if (!buffer_read_u8(cdata, &tmp)) THROW(SW_WRONG_DATA_LENGTH);
+
+    G_context.tx_info.has_change_output = (bool) tmp & 0x80;
```

Comment:
Add a comment explaining that the bit `0x80` indicates whether there's a change output or not.